### PR TITLE
Do not preallocate HirIds

### PR DESCRIPTION
**Pull request by cjgillot**
Part of https://github.com/rust-lang/rust/pull/87234

r? @petrochenkov

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
(Perf is now ready)

**Comment by petrochenkov** [comment]
Aha, here's the `style-servo-check` regression.

**Comment by cjgillot** [comment]
This PR also creates a `Found unstable fingerprints for opt_def_kind` ICE in the incr-patched cherry-pick case. I need to investigate it before this is merged.

**Comment by petrochenkov** [comment]
@cjgillot 
Could you set the `S-waiting-on-review` label once this is ready?
I see that you are updating the lowering related PRs from time to time, but I'm not sure whether they are finished or not, so I didn't take any action.

**Comment by cjgillot** [comment]
With the latest perfect results, I believe this one is ready.

File: compiler/rustc_ast_lowering/src/item.rs [hunk 1]
```diff
@@ -928,7 +920,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let has_value = true;
         let (defaultness, _) = self.lower_defaultness(i.kind.defaultness(), has_value);
         hir::ImplItemRef {
-            id: hir::ImplItemId { def_id: self.lower_node_id(i.id).expect_owner() },
+            id: hir::ImplItemId { def_id: self.allocate_hir_id_counter(i.id) },
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/item.rs:923]: I'm not sure why this difference between `lower_trait_item_ref` and `lower_impl_item_ref` exists.
For both of them `allocate_hir_id_counter` should have already been called by `visit_assoc_item`?

Comment (cjgillot) [reply compiler/rustc_ast_lowering/src/item.rs:923]: `lower_impl_item_ref` is called as part of `lower_item` for the impl block. At that time, `allocate_hir_id_counter` has not been called yet for the impl_item. However, this `allocate_hir_id_counter` is required for the call to `lower_visibility`. This difference is removed by https://github.com/rust-lang/rust/pull/87234 as it removes the call to ``lower_visibility`.

**Pull request merged**